egs/librispeech/ASR/pruned_transducer_stateless2/beam_search.py

```diff
@@ -624,6 +669,105 @@ def greedy_search_batch(
     for i in range(N):
         ans.append(sorted_ans[unsorted_indices[i]])
 
+    if not ngram_rescoring:
+        return ans
+
+    assert decoding_graph is not None
+
+    # Transform logits to shape [N, T, vocab_size] format to make it easier
+    # to select non-blank frames.
+    packed_all_logits = PackedSequence(
+        all_logits, torch.tensor(batch_size_list)
+    )
+    all_logits_unpacked, _ = pad_packed_sequence(
+        packed_all_logits, batch_first=True
+    )
+
+    # Transform non_blank_flag to shape [N, T]
+    packed_non_blank_flag = PackedSequence(
+        non_blank_flag, torch.tensor(batch_size_list)
+    )
+    non_blank_flag_unpacked, _ = pad_packed_sequence(
+        packed_non_blank_flag, batch_first=True
+    )
+
+    non_blank_logits_lens = torch.sum(non_blank_flag_unpacked, dim=1)
+    max_frame_to_rescore = non_blank_logits_lens.max()
+
+    non_blank_logits = torch.zeros(
+        [N, int(max_frame_to_rescore), vocab_size], device=device
+    )
+
+    # torch.index_select only acceptec a single dimension to index from.
+    # So we need generate non_blank_logits one by one.
+    # Maybe there is another efficient way to do this.
+    for i in range(N):
+        cur_non_blank_index = torch.where(non_blank_flag_unpacked[i, :] != 0)[0]
+        assert non_blank_logits_lens[i] == cur_non_blank_index.shape[0]
+        non_blank_logits[
+            i, : int(non_blank_logits_lens[i]), :
+        ] = torch.index_select(
+            all_logits_unpacked[i, :], 0, cur_non_blank_index
+        )
+
+    def verify_non_blank_logits():
+        # A way to verify non_blank_logits are selected correctly from all_logits.
+        hyps_before_rescore = non_blank_logits.argmax(dim=2)
+        for i in range(N):
+            usi = unsorted_indices[i]
+            hyp_to_verify = hyps_before_rescore[usi][
+                : int(non_blank_logits_lens[usi])
+            ].tolist()
+            assert ans[i] == hyp_to_verify
+        logging.info("Verified non-blank logits.")
+
+    # TODO: skip verification after we finally get a workable rescoring method.
+    if selection_verification:
+        verify_non_blank_logits()
+
+    # Split log_softmax into two seperate steps,
+    # so we cound do blank deweight in probability domain if needed.
+    logits_to_rescore_softmax = non_blank_logits.softmax(dim=2)
+    logits_to_rescore = logits_to_rescore_softmax.log()
+
+    # In paper: https://arxiv.org/pdf/2101.06856.pdf
+    # blank deweight is applied before non_blank frames selected.
+    # However, in current setup, that results in a higher WER.
+    # So just put this blank deweight before ngram rescoring.
+    # (TODO): debug this blank deweight issue.
+
+    blank_deweight = 100
```

Comment:
We previously tuned deweight value from 1-3, and could observe the deletion errors improvement. 100 may be too large?

Reply:
with current decoding method and 3-gram, blank_deweight only has tiny contribution, I tried [0, 10, 50, 100] and results are the same.

Maybe the bottleneck is the 3-gram or decoding method.

Reply:
>  I tried [0, 10, 50, 100] and results are the same.


It seems a problem of decoding_graph. gamma_blank works as expected after adding self loop to L.